`CloudRunWorker`

jakekaplan opened the pull request:
Adds the `CloudRuWorker`

Functionality has been taken from the `CloudRunJob` infrastructure block and ported over to the `Worker` paradigm.

## Example
```
pip install prefect-gcp
```

Start a Cloud Run Job Worker
```
prefect worker start -p 'my-work-pool' --type cloud-run
```

By default the `CloudRunWorker` will use a base jobs body with placeholders. These are populated from `CloudRunWorkerVariables`. 

```json
{
        "apiVersion": "run.googleapis.com/v1",
        "kind": "Job",
        "metadata": {
            "name": "{{ name }}",
            "annotations": {
                "run.googleapis.com/launch-stage": "BETA",
                "run.googleapis.com/vpc-access-connector": "{{ vpc_connector_name }}",
            },
        },
        "spec": {
            "template": { 
                "spec": {
                    "template": {
                        "spec": {
                            "containers": [
                                {
                                    "image": "{{ image }}",
                                    "args": "{{ args }}",
                                    "resources": {
                                        "limits": {
                                            "cpu": "{{ cpu }}",
                                            "memory": "{{ memory }}",
                                        },
                                        "requests": {
                                            "cpu": "{{ cpu }}",
                                            "memory": "{{ memory }}",
                                        },
                                    },
                                }
                            ],
                            "timeoutSeconds": "{{ timeout }}",
                            "serviceAccountName": "{{ service_account_name }}",
                        }
                    },
                },
            },
        },
    }
```

Similar to other workers using this base templating pattern allows us to provide a convenient way to add specific placeholders, as well as letting more advanced customization by modifying the actual template on a per work pool basis.

jakekaplan commented on prefect_gcp/worker.py:
```diff
@@ -0,0 +1,642 @@
+"""
+TODO docstring!
```
Working on this...

jakekaplan commented on prefect_gcp/worker.py:
```diff
@@ -0,0 +1,642 @@
+        default=None,
+        title="Memory",
+        description="The amount of memory allocated to the Cloud Run Job. See"
+        "https://cloud.google.com/run/docs/configuring/memory-limits#setting "
```
The `CloudRunJob` block had cpu and memory as `ints` as well as a memory units field. 

I opted to convert this to the actual value strings that should be used, but I can swap this back if anyone feels strongly

desertaxle replied:
I'm fine with either, but I lean more towards your implementation to have one field each for memory and CPU vs. two fields for each.

jakekaplan commented on prefect_gcp/worker.py:
```diff
@@ -0,0 +1,642 @@
+            ):
+                self.job_body["spec"]["template"]["spec"]["template"]["spec"][
+                    "containers"
+                ][0]["image"] = f"docker.io/{get_prefect_image_name()}"
```
running from my laptop gives `docker.io/prefecthq/prefect:dev-python3.10` which doesn't work on CloudRun, I'm not sure if we should force this to `docker.io/prefecthq/prefect:2-latest`?

madkinsz commented on prefect_gcp/worker.py:
```diff
@@ -0,0 +1,646 @@
+        "Compute Engine Service Account. ",
+        example="service-account@example.iam.gserviceaccount.com",
+    )
+    args: Optional[List[str]] = Field(
```
I think shlex is working well over in ECS

jakekaplan commented on tests/test_worker.py:
```diff
@@ -0,0 +1,574 @@
+        ]
+
+
+class TestCloudRunWorkerVariables:
```
Ahhhh, that makes sense. I will try to change to this tests with 
`deployment.check_valid_configuration(work_pool.base_job_template)`

desertaxle commented on prefect_gcp/worker.py:
```diff
@@ -0,0 +1,667 @@
+        description="The region where the Cloud Run Job resides.",
+        example="us-central1",
+    )
+    args: Optional[str] = Field(
```
@jakekaplan and I spoke offline about removing `args` and using only `command` for the default template. Leaving a note here so that I don't forget 😄

madkinsz commented on tests/test_worker.py:
```diff
@@ -556,19 +594,20 @@ def test_behavior_called_when_execution_get_fails(
+            with pytest.raises(Exception):
+                await cloud_run_worker.run(flow_run, cloud_run_worker_job_config)
+            calls = list_mock_calls(mock_client, 2)
+            # breakpoint()
```
:)

jakekaplan replied:
this is a faithful recreation of the test I copied over! (good catch, thank you 🙏)

jakekaplan commented on prefect_gcp/worker.py:
```diff
@@ -0,0 +1,659 @@
+    base job template.
+    """
+
+    region: str = Field(
```
@desertaxle is it incorrect for `region` to be a required value since there is no way to set it via `prefect work-pool create my-second-pool --type cloud-run`?

I left this as required because the infrablock had it required. But I don't see any other required fields on any other Worker setups

desertaxle replied:
I'd recommend adding a default for region to remove that friction. I think it should be fine to default to `us-central1` or a similar region.

jakekaplan commented on prefect_gcp/worker.py:
```diff
@@ -0,0 +1,836 @@
+"""
+<span class="badge-api beta"/>
+
+Module containing the Cloud Run worker used for executing flow runs as Cloud Run jobs.
```
A little unsure how to add this to the navigation and also preview this...

desertaxle replied:
I previewed it an it looks good to me!

jakekaplan commented:
This should be good to go re:
- addressing pr comments
- adding kill_infra methods

Things I'm unsure about:
- docs? do they get auto added? can I preview them?
- https://github.com/PrefectHQ/prefect-gcp/pull/172#discussion_r1171925298

desertaxle commented:
To preview docs you can run `mkdocs serve` locally. You might need to add some config to `mkdocs.yaml` for the documentation for the new module to show up.

desertaxle reviewed (approved):
LGTM!

kyirong6 commented:
Hey @jakekaplan , @desertaxle , @zanieb - I believe the default yaml file here needs to be updated in order to match googles change for the location of the `vpc-access-connector` annotation. Correct me if I'm wrong but the tag should be moved to the template's metadata. 

Here's the documentation I'm referencing: [VPC with Connectors](https://cloud.google.com/run/docs/configuring/vpc-connectors#configuring-job)

I ran into this error when filling the `VPC Connector Name` variable as a part of the Google Cloud Run Work Pool which led to me dig into this:

`googleapiclient.errors.HttpError: <HttpError 400 when requesting https://northamerica-northeast2-run.googleapis.com/apis/run.googleapis.com/v1/namespaces/bettercartanalytics/jobs?alt=json returned "metadata.annotations[run.googleapis.com/vpc-access-connector]: Annotation 'run.googleapis.com/vpc-access-connector' is not supported on resources of kind 'Service'. Supported kinds are: Revision, Execution". Details: "[{'@type': 'type.googleapis.com/google.rpc.BadRequest', 'fieldViolations': [{'field': 'metadata.annotations[run.googleapis.com/vpc-access-connector]', 'description': "Annotation 'run.googleapis.com/vpc-access-connector' is not supported on resources of kind 'Service'. Supported kinds are: Revision, Execution"}]}]">`

I changed the yaml file accordingly in the advanced tab which resolved the error:

```
{
  "job_configuration": {
    "command": "{{ command }}",
    "env": "{{ env }}",
    "labels": "{{ labels }}",
    "name": "{{ name }}",
    "region": "{{ region }}",
    "credentials": "{{ credentials }}",
    "job_body": {
      "apiVersion": "run.googleapis.com/v1",
      "kind": "Job",
      "metadata": {
        "name": "{{ name }}",
        "annotations": {
          "run.googleapis.com/launch-stage": "BETA"
        }
      },
      "spec": {
        "template": {
          "metadata": {
            "annotations": {
              "run.googleapis.com/vpc-access-connector": "{{ vpc_connector_name }}"
            }
          },
   ...
}
```

I'm currently using Prefect version 2.13.5 . Let me know if this is correct or not as I may have made an error elsewhere- thanks!
